### PR TITLE
MM-24634 Style emojis in sidebar with full opacity

### PR DESCRIPTION
**Pull request by hmhealey**
Since we use transparent text in many places to generate different colours, that has the side effect of making emojis appear faded. This adds a function to identify emojis in a string so that we can target them with CSS and style them with full opacity.

@andrewbrown00 For this, I made emojis full opacity for both active and inactive channels, but I left them faded for muted channels. They look the same when the channel is active/inactive, but I think that's fine since the surrounding text and background will still change colour. For muted channels, I think we still want them faded since it makes them less eye-catching. Let me know how this looks in the screenshots below.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-24634

#### Screenshots

Before: 
Screen Shot 2020-08-19 at 12 03 00 PM

After: 
Screen Shot 2020-08-19 at 11 57 43 AM

### CODE REVIEW

**Review by Willyfrog — APPROVED** [review]
LGTM a couple of minor non-blocking comments

File: utils/emoji_utils.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import emojiRegex from 'emoji-regex';
+import React from 'react';
+
+import {Emoji} from 'mattermost-redux/types/emojis';
+
+const defaultRule = (aName: string, bName: string, emojiA: Emoji, emojiB: Emoji) => {
+    if (emojiA && emojiB && emojiA.category === 'custom' && emojiB.category !== 'custom') {
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply utils/emoji_utils.tsx:10]: This came over when i migrated the file to TypeScript. I'm not sure why here actually, but I'm pretty sure that can never be the case. TypeScript doesn't think those can be null, so I think that should be a good enough excuse to remove them.

File: utils/emoji_utils.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    return emoji.aliases[0];
+};
+
+export function compareEmojis(emojiA: Emoji, emojiB: Emoji, searchedName: string) {
+    const aName = getEmojiName(emojiA, searchedName);
+    const bName = getEmojiName(emojiB, searchedName);
+
+    // Have the emojis that contain the search appear first
+    const aPrefix = aName.startsWith(searchedName);
+    const bPrefix = bName.startsWith(searchedName);
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply utils/emoji_utils.tsx:68]: I think the comment is the wrong one. I seem to remember something about the`:rice:` emoji being suggested when you typed `:ice`, so it got changed at some point.

**Review by sudheerDev — APPROVED** [review]

**Comment by andrewbrown00** [comment]
> My only concern seeing the screenshot side-by-side is we may want to let this change marinate a bit on daily to see if the emojis are to prominent in the category header or when a channel is unread and unselected, just a hunch. I'm 0/5 it will be an issue since I don't see emojis be over used.

**Update** I added a test server and played around with the emojis and they feel good. We'll see if any feedback comes in while this soaks on daily.

**Comment by hmhealey** [comment]
Yeah, I assume that either way, it’ll need a while for us to see how it works out. I don’t think we use emojis in names of things, but I’m sure someone does.

**Comment by jgilliam17** [comment]
Thanks @hmhealey 
Looks good. One small issue
- Some emoji still display with lower opacity, (see `:beach_umbrella: ` one) while other don't. I verified that being a lone emoji in the name is not an issue, it displays at lower opacity even with additional text, before or after emoji in the name. Also the same emoji remains at lower opacity regardless of `new` badge presence. 
Screen Shot 2020-08-26 at 9 23 02 AM

**Comment by hmhealey** [comment]
I can fix that case, but it will unfortunately make it so that some other emoji-like characters also always appear at full opacity in the sidebar. Unfortunately, we have no control over certain characters when the OS/browser decides to render them as a full colour emoji or as a black and white pictograph. In this case, the 🏖️ emoji is specified as a black and white text emoji, but Mac still renders it as a full colour one, so it's not recognized by the regex we're currently using. If I switch it to a regex that does capture text presentation emojis, then those will be at full opacity in the sidebar.

I think this way works ends up looking slightly better since I think it's more likely that someone will use 🏝️ than one of the more obscure ones that will cause that issue.

**Comment by hmhealey** [comment]
I spoke too soon. Those "emoji-like" characters I was talking about include plain old numbers. In this example, the numbers, copyright symbol, desert island emoji, and skull and crossbones emoji are "text presentation" emojis, so they appear faded in the PR branch
Screen Shot 2020-08-27 at 4 11 42 PM
and full brightness if I switch to the other emoji regex
Screen Shot 2020-08-27 at 4 11 18 PM

and we have no way to fix that without a lot of work to customize the 

@jgilliam17 I think we can just accept that this is a known issue and address it if anyone else brings it up. I don't think a fix is worth the effort at this time.

**Pull request merged**